rocksdb_replicator/replicated_db.cpp

```diff
@@ -139,11 +150,35 @@ rocksdb::Status RocksDBReplicator::ReplicatedDB::Write(
     // TODO(bol): This potentially could block all worker threads. We may
     // consider having a dedicated set of worker threads for admin requests,
     // and/or provide async write API when this turns out to be a problem.
-    if (!max_seq_no_acked_.wait(cur_seq_no, FLAGS_replicator_timeout_ms)) {
+    if (!max_seq_no_acked_.wait(cur_seq_no, current_replicator_timeout_ms_.load())) {
       incCounter(kReplicatorWriteWaitTimedOut, 1, db_name_);
       LOG(ERROR) << "Failed to receive ack from follower, timing out for " << db_name_;
+      numConsecutiveReplTimeout_ ++;
+
+      // enter degradation mode if there has been consecutive timeouts waiting for the follower ACK.
+      // This allows us to use a smaller wait timeout to fail fast, instead of blocking for long.
+      if (numConsecutiveReplTimeout_.load() == FLAGS_replicator_consecutive_ack_timeout_before_degradation
+          && current_replicator_timeout_ms_.load() == FLAGS_replicator_timeout_ms
+          && FLAGS_replicator_timeout_degraded_ms < FLAGS_replicator_timeout_ms
+          && FLAGS_replicator_timeout_degraded_ms >= kMinReplTimeoutMs) {
+        current_replicator_timeout_ms_.store(FLAGS_replicator_timeout_degraded_ms);
```

Comment:
These statements could be executed multiple times for a db when it is entering the degraded mode (from multiple threads) since you are not enclosing this in a mutex. I thought about it, and I think it should be safe. Just wanted to call it out in case you had other ideas.

Reply:
yeah, thanks for calling this out.  I was reluctant to use a mutex lock mainly for perf reasons. I also think it's okay, if two threads check these conditions at exactly the same time, they may decide to put the db in degradation mode by storing:
```
current_replicator_timeout_ms_.store(FLAGS_replicator_timeout_degraded_ms)
```
And that's idempotent. 
I can add a note for this.

Similarly for the below `current_replicator_timeout_ms_` check, I actually have a TODO for the same reason, but I thought that's just an optimization.